kubespawner/spawner.py

```diff
@@ -183,17 +187,25 @@ def __init__(self, *args, **kwargs):
         # By now, all the traitlets have been set, so we can use them to
         # compute other attributes
 
+        # namespace, pod_name, dns_name are persisted in state
+        # these same assignments should match clear_state
         if self.enable_user_namespaces:
-            self.namespace = self._expand_user_properties(self.user_namespace_template)
+            self.namespace = self._expand_user_properties(
+                self.user_namespace_template, slug_scheme="safe"
+            )
```

Comment:
I've been thinking about if/why slug_scheme should be hardcoded to "safe" here, leaving this comment for myself to not forget this open question.

Reply:
The `user_namespace_template` default is `"{hubnamespace}-{username}"`, so expanding that with `slug_scheme="safe"` in a hardcoded way here ought to hardcode a breaking change that will make PVCs be lost as they are namespaced resources.

I think if we hardcode `slug_scheme="safe"` here, we'd introduce a breaking change that can't be avoided by configuration for users of `enable_user_namespaces` that also use PVCs.

Reply:
I believe this is not _intended_ to be a breaking change because namespace is persisted in Spawner state. So this hardcoded scheme only affects _new_ spawners that have never launched before, and ensures namespace is always valid which the escape scheme does not. Otherwise, `namespace` is sticky and remembered for all launches after the first for a given Spawner.

I've updated the comment ot try to clarify the relationship of values set in `get_state` vs `__init__`.

Reply:
I think this may be a remnant of 'safe' not being the default. Now that it is, since escape is explicitly opt-in, I don't think we need to do this. I'll remove the hardcoded `safe` here, and leave it only on the labels, where it is actually required.